orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py

```diff
@@ -31,28 +31,39 @@ def _run_forward(execution_session, onnx_model, device, *inputs, **kwargs):
 
     # TODO: Try to reuse the output buffers as some of the output tensors are same sizes,
     #   especially the backward graph outputs.
-    io_binding = execution_session.io_binding()
-    run_options = C.RunOptions()
-
-    # Use IO binding
-    _utils._create_iobinding(io_binding, inputs, onnx_model, device)
-
-    # Run and return module outputs.
-    ort_output = execution_session.run_forward(io_binding, run_options)
-    forward_outputs, run_id = ort_output.ortvalues, ort_output.run_id
-
-    user_outputs = tuple(_utils._ortvalue_to_torch_tensor(forward_output) for forward_output in forward_outputs)
+    # REVIEW(codemzs): Consolidate Training Agent with InferenceAgent on C++ side to not
+    # have the need for passing IOBinding.
+    if isinstance(execution_session, onnxruntime.training.InferenceAgent):
+        io_binding = execution_session.io_binding()
```

Comment:
We have created `GraphExecutionManager`, `InferenceManager(GraphExecutionManager)` and `TrainingManager(GraphExecutionManager)` to separate inference and training through inheritance as opposed to if-else branches.

This change violates this principle. I'd recommend promoting `_run_forward` from free-style function to GraphExecutionManager method in which InferenceManager/TrainingManager could implement specificities for each training mode

Reply:
I agree with this, but I'm fine with it happening in a later PR.

It would make sense to stop using `IOBinding` in inference (as covered by a TODO somewhere else), in which case the two paths will probably mostly converge again!